resolver/resolver.cc

```diff
@@ -2299,30 +2300,38 @@ ast::ParsedFilesOrCancelled Resolver::run(core::GlobalState &gs, vector<ast::Par
     if (epochManager.wasTypecheckingCanceled()) {
         return ast::ParsedFilesOrCancelled();
     }
-    trees = resolveSigs(gs, std::move(trees));
-    if (epochManager.wasTypecheckingCanceled()) {
-        return ast::ParsedFilesOrCancelled();
+    result = resolveSigs(gs, std::move(trees));
+    if (!result.hasResult()) {
+        return result;
     }
-    sanityCheck(gs, trees);
+    sanityCheck(gs, result.result());
 
-    return ast::ParsedFilesOrCancelled(move(trees));
+    return result;
 }
 
-vector<ast::ParsedFile> Resolver::resolveSigs(core::GlobalState &gs, vector<ast::ParsedFile> trees) {
+ast::ParsedFilesOrCancelled Resolver::resolveSigs(core::GlobalState &gs, vector<ast::ParsedFile> trees) {
     ResolveSignaturesWalk sigs;
+    const auto &epochManager = gs.epochManager;
     Timer timeit(gs.tracer(), "resolver.sigs_vars_and_flatten");
     for (auto &tree : trees) {
+        if (epochManager->wasTypecheckingCanceled()) {
+            return ast::ParsedFilesOrCancelled();
+        }
```

Comment:
The last time we did this, we delayed the check once every 250 iterations:

https://github.com/sorbet/sorbet/pull/2654/files#diff-08da056e63e41cc7fbb183444db94a6cL296-L299

because wasTypecheckingCanceled uses a synchronization primitive, which is slow.

Thoughts on using something like that here?

Reply:
Good question. Let me measure overhead here and get back to you.

Reply:
I measured overhead on pay-server, and there's no noticeable increase in the runtime of any of the changed resolver passes. I think it's ~free if there's no contention on the atomics.